File: 2552-maximum-sum-of-distinct-subarrays-with-length-k/maximum-sum-of-distinct-subarrays-with-length-k.java
class Solution {
    public long maximumSubarraySum(int[] nums, int k) {
        long max = 0;
        long prefixSum = 0;
        
        // HashMap to store the frequency of elements in the current window
        HashMap<Integer, Integer> frequencyMap = new HashMap<>();
        
        // Initialize the first window of size k
        for (int i = 0; i < k; i++) {
            int curr = nums[i];
            frequencyMap.put(curr, frequencyMap.getOrDefault(curr, 0) + 1);
            prefixSum += curr;
        }
        
        // If the window contains exactly k unique elements, we update max
        if (frequencyMap.size() == k) {
            max = Math.max(max, prefixSum);
        }
        
        // Slide the window across the array
        for (int i = k; i < nums.length; i++) {
            int left = nums[i - k];
            int right = nums[i];
            
            // Remove the element that is leaving the window
            frequencyMap.put(left, frequencyMap.get(left) - 1);
            if (frequencyMap.get(left) == 0) {
                frequencyMap.remove(left);
            }
            prefixSum -= left;

            // Add the new element that is entering the window
            frequencyMap.put(right, frequencyMap.getOrDefault(right, 0) + 1);
            prefixSum += right;

            // If the window contains exactly k unique elements, check the sum
            if (frequencyMap.size() == k) {
                max = Math.max(max, prefixSum);
            }
        }
        
        return max;
    }
}

File: 2552-maximum-sum-of-distinct-subarrays-with-length-k/maximum-sum-of-distinct-subarrays-with-length-k.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number}
 */
var maximumSubarraySum = function (nums, k) {
    const freqMap = new Map();
    let maxSum = 0;
    let currSum = 0;

    for (let i = 0; i < k; i++) {
        const num = nums[i];
        freqMap.set(num, (freqMap.get(num) || 0) + 1);
        currSum += num;
    }

    if (freqMap.size === k) maxSum = Math.max(maxSum, currSum);

    for (let i = k; i < nums.length; i++) {
        const num = nums[i];
        freqMap.set(num, (freqMap.get(num) || 0) + 1);
        currSum += num;

        const old = nums[i - k];
        freqMap.set(old, (freqMap.get(old) || 0) - 1);
        if (freqMap.get(old) === 0) freqMap.delete(old);
        currSum -= old;

        if (freqMap.size === k) maxSum = Math.max(maxSum, currSum);
    }

    return maxSum;
};